### PR TITLE
allow custom types for `.svg` image imports (instead of default `any`)

### PR DESCRIPTION
**Pull request by stefanprobst**
currently, the built-in image import types referenced in `next-env.d.ts` type `.svg` imports as `any` (see [here](https://github.com/vercel/next.js/blob/canary/packages/next/image-types/global.d.ts#L10-L19)):

```ts
declare module '*.svg' {
  /**
   * Use `any` to avoid conflicts with
   * `@svgr/webpack` plugin or
   * `babel-plugin-inline-react-svg` plugin.
   */
  const content: any

  export default content
}
```

this pr adds a new `customImageTypes` experimental option - as suggested by @ijjk [here](https://github.com/vercel/next.js/pull/30060#pullrequestreview-1064489359) - which disables the built-in image import types, and allows users to provide their own types (via e.g. `custom-image-types.d.ts`)

x-ref: https://github.com/vercel/next.js/pull/30060

### CODE REVIEW

**Comment by ijjk** [comment]
Allow CI Workflow Run

- [ ] approve CI run for commit: 82e64df93c8e4ad73090823fd5dfec7134c909b4

Note: this should only be enabled once the PR is ready to go and can only be enabled by a maintainer

**Comment by ijjk** [comment]
Allow CI Workflow Run

- [ ] approve CI run for commit: 82e64df93c8e4ad73090823fd5dfec7134c909b4

Note: this should only be enabled once the PR is ready to go and can only be enabled by a maintainer

File: packages/next/src/lib/verify-typescript-setup.ts [hunk 1]
```diff
@@ -129,7 +131,7 @@ export async function verifyTypeScriptSetup({
     // Next.js' types:
     await writeAppTypeDeclarations({
       baseDir: dir,
-      imageImportsEnabled: !disableStaticImages,
+      imageImportsEnabled: !disableStaticImages && !customImageTypes,
```
Comment (styfle) [inline packages/next/src/lib/verify-typescript-setup.ts:134]: What is the difference between `customImageTypes` and `disableStaticImages`. Why not use the existing `disableStaticImages` instead?

**Review by styfle — CHANGES_REQUESTED** [review]
I don't think you need a config option to override types.

Instead, you can defined custom types.

1. Add a new file like `additional.d.ts`:

```ts
declare module '*.svg' {
  const content: { foo: string, bar: number }
  export default content
}
```

2. Then update tsconfig.json to include it:

```js
  "include": [
    "additional.d.ts",
    "next-env.d.ts",
    ".next/types/**/*.ts",
    "**/*.ts",
    "**/*.tsx",
  ],
```

See my previous comment here: https://github.com/vercel/next.js/pull/30060#discussion_r732243330

**Comment by stefanprobst** [comment]
@styfle unfortunately, this won't work - see https://github.com/stefanprobst/issue-next-svg-types/commit/e318c80b94035d9d90c0219634770fa9a7d09986 for a reproduction

Screenshot_20240103_232026

**Comment by styfle** [comment]
@stefanprobst Thats odd. I just tried with the latest `next@14.0.4` and it worked.

I had to put `additional.d.ts` as the first include in tsconfig.json, like this:

```json
{
  "compilerOptions": {
    "baseUrl": ".",
    "lib": [
      "dom",
      "dom.iterable",
      "esnext"
    ],
    "allowJs": true,
    "skipLibCheck": true,
    "strict": true,
    "noUncheckedIndexedAccess": true,
    "forceConsistentCasingInFileNames": true,
    "noEmit": true,
    "incremental": true,
    "esModuleInterop": true,
    "module": "esnext",
    "moduleResolution": "node",
    "resolveJsonModule": true,
    "isolatedModules": true,
    "jsx": "preserve",
    "plugins": [
      {
        "name": "next"
      }
    ]
  },
  "include": [
    "additional.d.ts",
    "next-env.d.ts",
    ".next/types/**/*.ts",
    "**/*.ts",
    "**/*.tsx",
  ],
  "exclude": [
    "node_modules"
  ]
}
```

**Comment by styfle** [comment]
Looks like it the difference is `skipLibCheck`. 

It works when true (which I believe is what TS recommends because its true when you run `tsc --init`).

**Comment by stefanprobst** [comment]
i set `skipLibCheck` to false in the issue repro, otherwise `.d.ts` files will not be typechecked (`skipLibCheck` also skips local `.d.ts` files).

with `skipLibCheck: true` i still only get `any` for `.svg` imports, not what is defined in `additional.d.ts`.

**Comment by styfle** [comment]
Thanks! I'll talk to the docs team to see how we can surface this information. It seems like this previous call-out was removed for some reason https://github.com/vercel/next.js/pull/30060#discussion_r732243330 so at the very least, we can add it back.

**Pull request closed without merging**